Unity-Technologies/ScriptableRenderPipeline
# Shader Graph to allow Dragging of Blackboard Properties accross different SGs windows

**Summary:**

* Fixes this bug: https://fogbugz.unity3d.com/f/cases/1199032/
* But also allows for Blackboard Properties to be dragged from one window to another (instead of just stopping the crash) by creating a new ShaderInput in the receiving graph & adding a node for it, because that was actually found to be an easy and straightforward solution to the bug, plus we get a mini feature out of it!

![DragAndDropAcrossGraphs0](https://user-images.githubusercontent.com/3393041/72401188-2e079d00-3700-11ea-9738-05aa29a8cd8e.gif)

---
**Implementation Notes:**

* Basically we are handed the other SG's ShaderInput in CreateNode, I just make a Copy of it and add it to the graph if the blackboard property doesn't appear to exist.
* No removed code, only new code.

---
**Manually Tested:**
* Various ShaderInputs (Properties & Keywords) with different types of settings (for example, if a copy a Vector1 from one graph to another and it's Mode is Slider, that will be copied over too).
* Moving something from graph 1 over to graph 2, changing some of the ShaderInputs's values in graph 2, then closing and reopening graph 1 and seeing if things were altered (to confirm that we aren't messing with the exact data: why the copy is necessary).
* Much dragging and dropping between graphs very fast.
* Dragging multiple things at once.

---
**Halo Effect:** 0/4 Only added code inside the method that handles dragging a new node into the graph from the blackboard
**Technical Risk:** 1/4

---
**Notes to QA:**

* As usual, see what I manually tested. I didn't exhaustively test each Property type and all things that could differ. But check that it is consistent with how duplicating ShaderInputs within the same graph works. For example "_Hybrid Instanced_" and "_Precision_" and "_Exposed_" aren't being duplicated, because that's how duplication is working within the same graph as well. With that said, I feel like some (all?) of those should be reported as separate bugs if they aren't already ("Duplication doesn't duplicate Precision").
**Update:** These have been reported.
* If you override a ShaderInput's Reference (such that it becomes bold), the moved over ShaderInput will also have that Reference Override. If you do not, the other SG will have a new "default" Reference. This is by design: if the reference matters to them then it will be overridden and as such copied over, otherwise a new reference should be made, but as always feel free to disagree if you don't think it should work that way.
* If you continue to move ShaderInputs that you have already dragged to anther graph, it will continue to make duplicates. This is by (slightly lazy) design: being smart by only adding the first instance and then afterwards just creating the node that uses that first instance of the ShaderInput is to support a very weird use case: this should be for copy ShaderInputs across graphs and nothing else. Making duplicates this way is perhaps what they want.
* If you drag and drop multiple blackboard properties at once, the new nodes will overlap. However this also happens if you do the same thing within the same graph. I suppose this should have a bugz case or favro card too.

---
**Notes to Documentation:**

Will add to Changelog in a bit, wanted to ask Sharlene something.

---
**Yamato:**
https://yamato.prd.cds.internal.unity3d.com/jobs/78-ScriptableRenderPipeline/tree/sg%252Fcopy-shader-input-across-graphs
